src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -78,20 +78,13 @@ private void StopRaisingEvents()
                 {
                     // Stop the RunLoop and wait for the thread to exit gracefully
                     Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
-                    _watcherThread.Join();
-                    _watcherRunLoop = IntPtr.Zero;
                 }
 
-                // Clean up the EventStream, if it exists
-                if (!_eventStream.IsInvalid)
-                {
-                    _eventStream = new SafeEventStreamHandle(IntPtr.Zero);
-                }             
-
-                // Cleanup the callback
-                if (_callback != null)
+                // Gurd against us from calling Join on our own background thread (for 
+                // example, calling Dispose or Stop from a callback).
+                if (Thread.CurrentThread != _watcherThread)
                 {
-                    _callback = null;
+                    _watcherThread.Join();
```

Comment:
Hmm good point; there is the potential for corrupt state here. I think the easiest fix would be the guard around the Thread.Join; that guard should prevent rapid-fire start-stop state corruption and shouldn't cause a terrible perf hit.

Thoughts?